libmachine/persist/secretstore.go

```diff
@@ -0,0 +1,203 @@
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &SecretStore{Store: store, SecretName: secretName, SecretNamespace: secretNamespace, SecretClient: clientset.CoreV1().Secrets(secretNamespace)}
+	if err := s.extractConfig(); err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+func (s *SecretStore) Remove(name string) error {
+	if err := s.Store.Remove(name); err != nil {
+		return fmt.Errorf("error removing directories for host %v: %v", name, err)
+	}
+	return s.saveSecret(name)
+}
+
+func (s *SecretStore) Save(host *host.Host) error {
+	if err := s.Store.Save(host); err != nil {
+		return fmt.Errorf("error saving with file store: %v", err)
+	}
+	return s.saveSecret(host.Name)
+}
+
+func (s *SecretStore) saveSecret(hostName string) error {
+	// create the tar.gz file
+	destFile := &bytes.Buffer{}
+
+	fileWriter := gzip.NewWriter(destFile)
+	tarfileWriter := tar.NewWriter(fileWriter)
+
+	if err := s.addDirToArchive(tarfileWriter); err != nil {
+		return err
+	}
+
+	tarfileWriter.Close()
+	fileWriter.Close()
+
+	secret, err := s.loadSecret()
+	if err != nil {
+		return fmt.Errorf("unable to load secret : %v", err)
+	}
+
+	if secret.Data == nil {
+		secret.Data = make(map[string][]byte)
+	}
+	secret.Data["extractedConfig"] = destFile.Bytes()
+
+	secret, err = s.SecretClient.Update(context.Background(), secret, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to update secret: %v", err)
+	}
+
+	return nil
+}
+
+func (s *SecretStore) addDirToArchive(tarfileWriter *tar.Writer) error {
+	baseDir := s.GetMachinesDir()
+
+	return filepath.Walk(baseDir,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			if path == baseDir || strings.HasSuffix(info.Name(), ".iso") ||
+				strings.HasSuffix(info.Name(), ".tar.gz") ||
+				strings.HasSuffix(info.Name(), ".vmdk") ||
+				strings.HasSuffix(info.Name(), ".img") {
+				return nil
+			}
+
+			header, err := tar.FileInfoHeader(info, info.Name())
+			if err != nil {
+				return err
+			}
+
+			header.Name = path
+
+			if err := tarfileWriter.WriteHeader(header); err != nil {
+				return err
+			}
+
+			if info.IsDir() {
+				return nil
+			}
+
+			file, err := os.Open(path)
+			if err != nil {
+				return err
+			}
+			defer file.Close()
+
+			_, err = io.Copy(tarfileWriter, file)
+			return err
+		})
+}
+
+func (s *SecretStore) loadSecret() (*v1.Secret, error) {
+	secret, err := s.SecretClient.Get(context.Background(), s.SecretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error getting secret from kubernetes: %v", err)
+	}
+
+	return secret, nil
+}
+
+func (s *SecretStore) extractConfig() error {
+	secret, err := s.loadSecret()
+	if err != nil {
+		return fmt.Errorf("error getting secret from kubernetes: %v", err)
+	}
+
+	baseDir := s.GetMachinesDir()
```

Comment:
nit: could move this next to `tarReader` since it isn't needed until for loop.

Reply:
Done.